Big upgrade

sauloperez opened the pull request:
Well, well, well. This is the long overdue upgrade of the Expo SDK. Beware, this is a pretty big PR.

I had to do a complete rewrite, but given how small the app is, it wasn't a huge deal. The Expo APIs have changed a lot, but most of all, we were light years apart from React looks like today.

## How?

As this is the first time I do anything with Expo, I learned and experimented first with https://github.com/sauloperez/my-app. I didn't move back to this repo until I got something more or less working there.

## Detailed list of changes

* This makes use of the new [EAS Build](https://docs.expo.dev/build/introduction/) to run builds. Check the details in the changed README.
* Configures the project to use Firebase Cloud Messaging to receive push notifications in Android. No idea how they worked before but it seems that a lot has changed here.

## Additions

Because my time is precious, I decided to add Prettier and ESLint with very basic configurations so they could ease my work a bit. They are, by no means, all we might need. I thought of adding them to a separate PR but I needed them for these changes, so I figured we can improve later on.


![](https://media.giphy.com/media/mi6DsSSNKDbUY/giphy-downsized-large.gif)

@enricostano @coopdevs/timeoverflow-contributors thoughts welcome.

sauloperez committed:
Keep track of node version

sauloperez committed:
Upgrade to SDK 33

sauloperez committed:
Upgrade to SDK 35

By simply running `expo update 35.0.0`.

sauloperez committed:
Upgrade to SDK 41

By running `expo upgrade`. This is the oldest supported version.

sauloperez committed:
Upgrade to SDK 43

By running `expo upgrade`. This is the oldest supported version by Expo
Go, the dev mobile app.

sauloperez committed:
Update install instructions in README

The current ones are long deprecated.

sauloperez committed:
Upgrade to Expo latest (46.0)

This is the result of running `expo upgrade`.

sauloperez committed:
Update the setup and dev steps

This is what I have followed so far, except the `npm install`. I haven't
installed packages yet.

sauloperez committed:
Update config settings and packages

I created a new project from scratch and copied over the values I got by
default. I kept the values that seemed to make sense.

Additionally, I updated the `versionCode`. It seems that every new build
uploaded to the respective platform (iOS or Android), requires a new
version. That's what I could understand from a `git blame` and
https://docs.expo.dev/eas-update/runtime-versions/. I'll forget about
this when I push a new build.

I applied the same logic for the project's version.

sauloperez committed:
Tear app apart and start from scratch

I will rewrite the features one by one.

sauloperez committed:
Update .gitignore

Taken from a vanilla Expo project.

sauloperez committed:
Load webview

sauloperez committed:
Visit different host depending on releaseChannel

A lot has changed in recent SDK versions in the way builds are published. The
way to go now is with EAS Update. See:
https://docs.expo.dev/eas-update/introduction/. This Expo's new build
system seems to ease the process of building and publishing to app
stores. See: https://docs.expo.dev/build/setup/.

This adds an EAS file as a result of following the steps listed in
https://docs.expo.dev/eas-update/getting-started/.

```
npm install -g eas-cli expo-cli
eas login # using Coopdevs Expo credentials
npx expo install expo-updates
eas update:configure
eas build:configure
npx expo start
```

Although the `Constants` package still exists and contains the
`releaseChannel` key
(https://docs.expo.dev/versions/latest/sdk/constants/#expoclientconfig),
it wasn't present while developing (we need to set it ourselves?). Their
docs at
https://github.com/expo/expo/blob/8682b075917a7e92b83e5e7ea7711f98eadfcb56/docs/pages/archive/classic-updates/release-channels.mdx#access-channel-from-code
and
https://docs.expo.dev/versions/latest/sdk/updates/#updatesreleasechannel
mention this has been moved to the `expo-updates` package.

All this setup will enable creating developments builds. See:
https://docs.expo.dev/development/build/.

sauloperez committed:
Implement back button behavior

sauloperez committed:
Implement push notifications

Note this also stores the device token in the backend. This ports
https://github.com/coopdevs/timeoverflow-mobile-app/pull/1.

sauloperez committed:
Replace then() with async/await

sauloperez committed:
Select "Remember me" option by default

Ports https://github.com/coopdevs/timeoverflow-mobile-app/pull/2.

sauloperez committed:
Open external links in a web browser

sauloperez committed:
Show status bar with same color as navbar

It shows a darker bottom border on the bar that I don't know how to
remove. I don't know which component causes it.

sauloperez committed:
Simplify config of base url the app points to

This decouples the code from the release channels. We only need to
specify the base url when making a build. I had to provide a default
because when using the Expo Go client in development there's no build. I
prefered this over requiring to pass an ENV var to `nxp expo start` or
setting up dir env. This is the scenario
https://docs.expo.dev/build-reference/variables/#can-i-share-environment-variables-defined-in
describes. Those solutions seemed too convoluted for our simple
duplication.

This makes the eas config look like a vanilla one, as stated in
https://docs.expo.dev/build/eas-json/#common-use-cases. This separates
the concept of "build profile" from "release channel", which let's us
combine them as we wish, without having to change the app's code. More
details about channels:
https://docs.expo.dev/eas-update/how-eas-update-works/#distributing-builds.

sauloperez committed:
Move functions back to lib/

sauloperez committed:
Upgrade Android's versionCode

This is meant for Play Store to distinguish between binaries.

sauloperez committed:
Consistently name build profiles and channels

According to
https://docs.expo.dev/eas-update/how-eas-update-works/#matching-updates-and-builds,
this will make sure "preview" builds end up in the app store's staging.
I have no clue how to check this though.

It avoids the need to get familiar with another concept; the build
profile. It's also highly unlikely that we have to combine different
profiles and channels.

sauloperez committed:
Upgrade versionCode to a potentially unused one

Sergi played with version codes in the past and we don't know which one
may be available.

sauloperez committed:
Configure Android's push notifications with FCM

This is the step described in
https://docs.expo.dev/push-notifications/using-fcm/#client-setup.

Note we also followed
https://docs.expo.dev/push-notifications/using-fcm/#uploading-server-credentials
and uploaded the credentials to our Expo account. We initially missed it
and pushes weren't working.

sauloperez committed:
Update README with new build command

A lot has changed in Expo with the introduction of EAS Build.

sauloperez committed:
Set up prettier + eslint with very basic config

I feel safer this way. I don't want to keep thinking whether I want
single or double quotes, and I want someone to point me out dumb
mistakes I may make in JS.

sauloperez commented on App.js:
```diff
@@ -1,66 +1,99 @@
+      } catch (err) {
+        console.log("[handleBackButtonPress] Error : ", err.message);
+      } finally {
+        return true; // eslint-disable-line no-unsafe-finally
```
I need to figure this out. I'm pretty sure we don't need to return anything.

markets commented:
Wow @sauloperez! Such an amazing work 👏

I have 0 experience with Expo, so I can't help so much with this 😅, but I made a quick review and everything seems very resonable 👌

markets commented on lib/injectCustomJavaScript.js:
```diff
@@ -1,19 +1,19 @@
 
 const handleLoginPage = (webview) => {
-  webview.injectJavaScript("$('#user_remember_me').prop('checked', true);")
+  webview?.injectJavaScript("$('#user_remember_me').prop('checked', true);");
```
@sauloperez maybe we can write this small js expression without jQuery, using vanilla js? So if we eventually remove jQuery from main app, this will continue to work.

sauloperez replied:
yeah, I think we can clearly do that in Vanilla JS. I'll look into it.

markets replied:
@sauloperez I think this change should be enough:

```suggestion
  webview?.injectJavaScript("document.querySelector('#user_remember_me').checked = true");
```

sauloperez replied:
yep, I think so. I'll try it out as soon as I can get back to this.

sseerrggii commented:
How we can test @sauloperez? In the past we only test android app

sauloperez commented:
I haven't tried yet to get an iOS build, but in the meantime, you can already start testing an Android one :point_right: https://expo.dev/accounts/coopdevs/projects/time-overflow/builds/9fd2d032-83ff-4c40-b92d-ebd79e2975b9.

I'll try to address @markets 's point in the meantime.

sseerrggii commented:
Tested ✅
Everything seems fine! 
Good job @sauloperez 😘

sauloperez commented:
There's something going on with the push notification's icon though. This is what I got from your testing
![Screenshot_20221121-131325](https://user-images.githubusercontent.com/762088/203052556-da7ba815-5b9c-4436-bc48-813885e64797.png)

sseerrggii commented:
With icon should appear @sauloperez ?

Not a big deal for us, we can continue without icon, but a icon of a sand clock or something similar will be good

sauloperez committed:
Remove dependency on jQuery

This stems from
https://github.com/coopdevs/timeoverflow-mobile-app/pull/13#discussion_r1021371419.

sauloperez committed:
Restore push notification icon

See https://docs.expo.dev/versions/latest/config/app/#notification for
reference. Apparently, this got lost in 42c2192.

sauloperez commented:
I just pushed another build in https://expo.dev/accounts/coopdevs/projects/time-overflow/builds/004bae94-deac-41eb-9fe9-a12ccfa30f79 to try out my last commit where the icon issue should've been fixed. Feel free to check it again, @sseerrggii

sauloperez commented:
can you check with your phone, @sseerrggii ? I still fail to see the icon

sauloperez commented:
shall we deploy this to production anyway? or do you want to investigate this icon issue in more depth?

sseerrggii commented:
Deploy. We will fix this later

Thanks @sauloperez

sauloperez commented:
I'm also having issues trying to sign and upload the app to the Play Store with Expo :see_no_evil: but we'll manage
